text/0000-codify-infrastructure.md

```diff
@@ -0,0 +1,392 @@
+    + [DJs and VJs 💽](#djs-and-vjs-%F0%9F%92%BD)
+    + [Developers 👩🏻‍💻](#developers-%F0%9F%91%A9%F0%9F%8F%BB%E2%80%8D%F0%9F%92%BB)
+    + [Management 🤵🏻](#management-%F0%9F%A4%B5%F0%9F%8F%BB)
+  * [Path to Production 👷🏻‍♀️](#path-to-production-%F0%9F%91%B7%F0%9F%8F%BB%E2%80%8D%E2%99%80%EF%B8%8F)
+- [Drawbacks 😬](#drawbacks-%F0%9F%98%AC)
+- [Alternatives ⚖️](#alternatives-%E2%9A%96%EF%B8%8F)
+- [Unresolved questions ❓](#unresolved-questions-%E2%9D%93)
+
+<!-- tocstop -->
+
+## Summary 📖
+
+Right now, the infrastructure for datafruits is really ad-hoc. As best as I can tell, some stuff is running on [Heroku](https://www.heroku.com/) and some stuff is running on [DigitalOcean](https://www.digitalocean.com/). This RFC has two goals:
+
+1. Codify the infrastructure for the datafruits site so that it's easily understandable where everything lives.
+2. Cloud-ify the existing infrastructure so that each component is individually scalable.
+
+## Motivation 🏃🏻‍♀️
+
+As stated above, one reason to do this is to make it more clear where everything lives -- it allows someone (like me ☺️) to take a look and understand how all the pieces fit together.
+
+Another reason to keep the infrastructure as code is that it makes the infra versionable, meaning rollbacks are literally as easy as `git revert`. [More on the benefits of IoC](https://docs.microsoft.com/en-us/azure/devops/learn/what-is-infrastructure-as-code)
+
+This also allows us to move from treating the services (streampusher, datafruits) like pets to treating them like cattle, which means less messy ssh-ing and more codified changes. [More reading on this terminology](https://medium.com/@Joachim8675309/devops-concepts-pets-vs-cattle-2380b5aab313)
+
+Additionally, by moving to a [Kubernetes (K8s)](https://kubernetes.io/) based infrastructure, it allows us a lot more (horizontal) scalability, and the ability to scale out each chunk of the stack individually -- if there are 30,000,000 users in the chat, it makes sense to scale up the chat server, but maybe not also the [rmtp(s)](https://en.wikipedia.org/wiki/Real-Time_Messaging_Protocol) server, which only has the one person running viz connected to it.
+
+Fourth, we now have the opportunity to actually do logging. Logging! Imagine, datafruits goes down and we don't have to spend eight hours figuring out why. We have an idea for a new feature, and we can actually figure out if people use it. Metrics! God, the metrics. Plus we can avoid G\*\*\*\*\* A\*\*\*\*\*\*\*\* and run our [own analytics platform](https://github.com/zgoat/goatcounter) that's privacy friendly.
+
+Lastly, there are also cost savings associated with moving to a containerized architecture. We move from running a few digital ocean droplets, a couple heroku dynos, and whatever else we have to running 3 droplets on digital ocean. Additionally, those droplets will be a lot more efficient about their resource usage than they currently are, since their only purpose will be to run docker images.
```

Comment:
Current costs are something like
1 digital ocean droplet for streampusher api/postgresql/redis/liquidsoap/main icecast -> $40
1 digital ocean droplet for icecast relay -> $5
heroku dyno for datafruits.fm w/ fastboot -> $7
heroku dyno for hotdoglounge chat -> $7

I think that's all...

Reply:
Just using some quick napkin math based on [Digital Ocean pricing](https://www.digitalocean.com/pricing/):

[Managed Postgres DB](https://www.digitalocean.com/pricing/#managed-databases):

Memory | vCPUs | Disk | Standby Nodes | $/HR | $/MO
-- | -- | -- | -- | -- | -- |
1GB | 1vCPU | 10GB | N/A | $0.022 | $15

[Spaces Object Storage](https://www.digitalocean.com/pricing/#spaces-object-storage):

Storage | Outbound Transfer | Add'l. GB Stored | Add'l. GB Transferred | $/MO 
-- | -- | -- | -- | --
250GB | 1TB | $0.02/GB | $0.01/GB | $5

[Standard Droplets](https://www.digitalocean.com/pricing/#standard-droplets):

Memory | vCPUs | Transfer | SSD Disk | $/HR | $/MO
-- | -- | -- | -- | -- | -- 
1GB | 1vCPU | 1TB | 25GB | $0.007 | $5

Just doing some quick napkin math:

Component | $/MO
-- | --
Droplet (x3) | $15
Postgres | $15
Object Storage | $5
**Total** | **$35**

I didn't realize we were using redis -- what's that being used for?

Reply:
> I didn't realize we were using redis -- what's that being used for?

Sidekiq (background worker) process in Rails, some other things in Rails, its also a way for Rails to communicate with the hotdog lounge chat app.

Reply:
Gotcha -- If that's the case we'd likely use what's called the [sidecar pattern](https://thenewstack.io/tutorial-apply-the-sidecar-pattern-to-deploy-redis-in-kubernetes/) where we'd have a Redis client container alongside both the hotdog_lounge containers and the Rails API container, and both of those would connect to a [managed redis instance](https://www.digitalocean.com/products/managed-databases-redis/). Adding that on would bring the total to $50/MO.

The sidecar pattern is also nice because for local development you can still use localhost, since the containers will share a pod.